framework/doc/content/source/runners/Runner.md

```diff
@@ -0,0 +1,184 @@
+# Runner
+
+The Runner controls the "execution" behavior of the simulation. They direct
+the flow of solves, stepping, output recording, etc.  Runners are similar to
+the original Executioner system in that they have one primary virtual function
+that each runner implements that defines its behavior.  However, runners
+differ in that they are designed to be composed into arbitrary tree structures
+both in-code and by users via input files - similar to how the MeshGenerator
+system works.  The Runner system is currently highly experimental and has not
+yet stabilized and been fully implemented.  For this reason, MOOSE will only
+run in this mode if given the `--runner` flag on the command line.  The system
+name may change, input syntax may change, etc.  YOU WERE WARNED!
+
+## Using Runners from input files
+
+When using runners, you must remember to pass in the `--runner` flag on the
+cli when you run your application binary.  This allows the Executioner block
+to be omitted and causes MOOSE to ignore it if present.  Support for a new
+`[Runner]` block has been added to input files.  Users will populate this
+block with equivalent content that normally was present in the `[Executioner]` block.
+A hypothetical example is this:
+
+```
+[Runner]
+  [solve]
+    type = FooSolver
+    max_its = 42
+    ...
+  []
+  [refine]
+    type = Refine
+    inner = solve
+  []
+  [init]
+    type = Init
+    inner = refine
+    petsc_options = ...
+  []
+[]
+```
+
+Which might perform the equivalent of something like the current `Steady`
+executioner.  In this example, "init" becomes the primary runner which does
+some things to set up and tear down the simulation.  Inside its setup and
+tear-down, it executes an inner runner that has been set to "refine".  The
+"refine" runner does some mesh refinement things and then executes an inner
+runner - which here has been set to "solve". This allows refinement to be
+wrapped/inserted into arbitrary areas of the simulation execution process.
+
+The new system, is also capable of providing alias-like runners that generate
+equivalent runner trees programmatically in order to replicate the current
+`Steady` or `Transient` behavior like users currently expect now:
+
+```
+[Runner]
+  [steady]
+    type = FauxSteady
+    solve_type = 'PJFNK'
+    petsc_options_iname = '-pc_type -pc_hypre_type'
+    petsc_options_value = 'hypre boomeramg'
+    ...
+  []
+[]
+```
+
+This runner would simply generate the init+solve+refine trio of runners
+programmatically - hiding the runner structure from the user.  By default the
+last runner listed in the `Runner` block becomes the master/primary runner.
```

Comment:
This shouldn't be necessary - we should find the root runner by sorting the graph using the dependency resolver.

Reply:
I agree.  The trick is that I'd like runners to be able to grab their child/sub runners in their constructors - but that only works if we create the objects themselves in the correct order.  But we can't run any normal dependency resolver interface or similar techniques on objects that haven't been created yet.  So I'm not sure the best approach.  Have we resolved this in e.g. mesh generators in any way that doesn't require inter-object binding to go outside of the constructor?